SWATCH-1937: Refactor contract ingestion to increase consistency

Sgitario opened the pull request:
Jira issue: [SWATCH-1937](https://issues.redhat.com/browse/SWATCH-1937)

## Description
- added upsertContract
- refactored POST contract endpoint
- updated README.md for swatch-contracts

Note for QE, the API for the contract creation has changed (the request and the response). See the changes in the README.md for more information about how to fullfill the new request and what to expect in the new response.

## Testing
1.- podman-compose up
2.- start our API service:
```
SERVER_PORT=8101 DEV_MODE=true SPRING_PROFILES_ACTIVE=api,kafka-queue ./gradlew :bootRun
```
3.- add the ROSA offering:

```
INSERT INTO public.offering (sku, product_name, product_family, cores, sockets, hypervisor_cores, hypervisor_sockets, role, sla, usage, description, has_unlimited_usage, derived_sku,metered) 
VALUES ('RH00001', 'OpenShift Online', '', null, 5, null, null, 'moa-hostedcontrolplane', 'Premium', '', 'Red Hat OpenShift Dedicated on Customer Cloud Subscription (Hourly)', false, 'Article_1589801907847034','true');
```

4.- start the swatch-contracts service:

```
SUBSCRIPTION_URL=http://localhost:8102 QUARKUS_MANAGEMENT_PORT=9001 ./gradlew :swatch-contracts:quarkusDev
```

5.- create contract:

```
curl -v -X POST http://localhost:8000/api/swatch-contracts/internal/contracts \
-H 'Content-Type: application/json' \
--data-binary @- << EOF
{
  "partner_entitlement": {
        "rhAccountId": "123456",
        "sourcePartner": "aws_marketplace",
        "entitlementDates": {
          "startDate": "2023-05-24T18:46:05Z",
          "endDate": "2023-05-24T20:46:06.273014Z"
        },
        "rhEntitlements": [
          {
            "subscriptionNumber": "12585274",
            "sku": "RH00001"
          }
        ],
        "purchase": {
          "vendorProductCode": "AAAA",
          "contracts": [
            {
              "planId": "<only for azure>",
              "dimensions": [
                {
                  "name": "four_vcpu_hour",
                  "value": "8"
                }
              ]
            }
          ]
        },
        "partnerIdentities": {
          "awsCustomerId": "BBB",
          "sellerAccountId": "CCC",
          "customerAwsAccountId": "DDD"
        }
  },
  "subscription_id": "123456456"
}
EOF
```

Output:
```
{
  "status": {
    "status": "SUCCESS",
    "message": "New contract created"
  },
  "contract": {
    "uuid": "97d53683-417e-4a8b-8451-f799c001c2c5",
    "subscription_number": "12585274",
    "sku": "RH00001",
    "start_date": "2023-05-24T18:46:05Z",
    "end_date": "2023-05-24T20:46:06.273014Z",
    "org_id": "123456",
    "billing_provider": "aws",
    "billing_provider_id": "AAAA;BBB;CCC",
    "billing_account_id": "DDD",
    "product_id": "rosa",
    "vendor_product_code": "ezcoaphi4bqc7lktoy2qfd6zi",
    "metrics": [
      {
        "metric_id": "four_vcpu_hour",
        "value": 8
      }
    ]
  }
}
```

awood commented on swatch-contracts/src/main/java/com/redhat/swatch/contract/service/AzurePartnerEntitlementsProvider.java:
```diff
@@ -0,0 +1,44 @@
+public class AzurePartnerEntitlementsProvider extends BasePartnerEntitlementsProvider {
+
+  @Override
+  public boolean isValid(PartnerEntitlementContract contract) {
+    return Objects.nonNull(contract.getCloudIdentifiers())
+        && Objects.nonNull(contract.getCloudIdentifiers().getAzureResourceId());
+  }
```
Same question about bean validation here.

awood commented on swatch-contracts/src/main/java/com/redhat/swatch/contract/service/AwsPartnerEntitlementsProvider.java:
```diff
@@ -0,0 +1,46 @@
+public class AwsPartnerEntitlementsProvider extends BasePartnerEntitlementsProvider {
+
+  @Override
+  public boolean isValid(PartnerEntitlementContract contract) {
+    return Objects.nonNull(contract.getRedHatSubscriptionNumber())
+        && Objects.nonNull(contract.getCloudIdentifiers())
+        && Objects.nonNull(contract.getCloudIdentifiers().getAwsCustomerId())
+        && Objects.nonNull(contract.getCloudIdentifiers().getAwsCustomerAccountId())
+        && Objects.nonNull(contract.getCloudIdentifiers().getProductCode());
+  }
```
Could we use bean validation here instead?  I bring it up because @kahowell just retired some very similar code in #3050 [here](https://github.com/RedHatInsights/rhsm-subscriptions/pull/3050/files#diff-d0a3f94bef59f416c81d8ea1f761cdfcd6e5ae3e5888e580e5b398772ce8b5f1L499-L509).  I realize it's a bit tricky since the `PartnerEntitlementContract` is validated in different ways depending on context, but maybe it's worth a little research into whether there's a solution with bean validation here.

Just taking a quick look, it seems like there might be away to apply different validation rules at [runtime](https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#_groupsequenceprovider).  Chapter 5 in that document seems broadly relevant.  We could have validation groups for `CloudIdentifiers`: one for Azure, one for AWS, and one default (which would be a `NonNull` on `getProductCode`) and then apply the correct groups based on a `GroupSequenceProvider` implementation.

kahowell replied:
https://quarkus.io/guides/validation#validation-groups-for-rest-endpoint-or-service-method-validation ?

Sgitario replied:
This is not a validation method, but a way to identify if it's for azure or aws. The `isValid` method name is misleading, will rename it.

kahowell commented on swatch-contracts/README.md:
````diff
@@ -9,18 +9,18 @@ If you want to learn more about Quarkus, please visit its website: https://quark
 
 ```shell script
-./gradlew quarkusDev
+../gradlew quarkusDev
````
Better to be explicit about project:
```suggestion
../gradlew :swatch-contracts:quarkusDev
```

Sgitario replied:
+1

awood commented:
And coincidentally enough, when I make the `curl` call listed, the following stack gets thrown

```
.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method                                                                                                                                                                                                              
        at com.redhat.swatch.contract.resource.ContractsTestingResource_Subclass.createContract(Unknown Source)                                                                                                                                                                           
        at com.redhat.swatch.contract.resource.ContractsTestingResource_ClientProxy.createContract(Unknown Source)                                                                                                                                                                        
        at com.redhat.swatch.contract.openapi.resource.DefaultApi$quarkusrestinvoker$createContract_1b3ee7b50158f76928c19d139536ec966e38a873.invoke(Unknown Source)                                                                                                                       
        at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)                                                                                                                                                                                
        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)                                                                                                                                
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: jakarta.transaction.RollbackException: ARJUNA016053: Could not commit transaction.
        at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1283)
        at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
        at io.quarkus.narayana.jta.runtime.NotifyingTransactionManager.commit(NotifyingTransactionManager.java:70)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.endTransaction(TransactionalInterceptorBase.java:406)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:171)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
        at io.quarkus.resteasy.reactive.server.runtime.StandardSecurityCheckInterceptor.intercept(StandardSecurityCheckInterceptor.java:44)
        at io.quarkus.resteasy.reactive.server.runtime.StandardSecurityCheckInterceptor_RolesAllowedInterceptor_Bean.intercept(Unknown Source)                                                                                                                                            
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
        at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
        ... 13 more                                                   
Caused by: jakarta.validation.ConstraintViolationException: Validation failed for classes [com.redhat.swatch.contract.repository.ContractEntity] during persist time for groups [jakarta.validation.groups.Default, ]
List of constraint violations:[                                       
        ConstraintViolationImpl{interpolatedMessage='must not be empty', propertyPath=metrics, rootBeanClass=class com.redhat.swatch.contract.repository.ContractEntity, messageTemplate='{jakarta.validation.constraints.NotEmpty.message}'}
]                                                                     
        at org.hibernate.boot.beanvalidation.BeanValidationEventListener.validate(BeanValidationEventListener.java:151)
        at org.hibernate.boot.beanvalidation.BeanValidationEventListener.onPreInsert(BeanValidationEventListener.java:81)
        at org.hibernate.action.internal.EntityInsertAction.preInsert(EntityInsertAction.java:215)
        at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:98)
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:635)
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:502)
        at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:364)
        at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
        at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1412)
        at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:485)
        at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2301)
        at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1966)
        at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439)
        at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:336)                                                                                                                                 
        at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
        at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
        at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:52)
        at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:351)
        at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:69)
        at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:138)
        at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1271)
        ... 30 more                                                   
```

Sgitario commented:
> And coincidentally enough, when I make the `curl` call listed, the following stack gets thrown
> 
> ```
> .quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method                                                                                                                                                                                                              
>         at com.redhat.swatch.contract.resource.ContractsTestingResource_Subclass.createContract(Unknown Source)                                                                                                                                                                           
>         at com.redhat.swatch.contract.resource.ContractsTestingResource_ClientProxy.createContract(Unknown Source)                                                                                                                                                                        
>         at com.redhat.swatch.contract.openapi.resource.DefaultApi$quarkusrestinvoker$createContract_1b3ee7b50158f76928c19d139536ec966e38a873.invoke(Unknown Source)                                                                                                                       
>         at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)                                                                                                                                                                                
>         at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)                                                                                                                                
>         at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
>         at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
>         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
>         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
>         at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
>         at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
>         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
>         at java.base/java.lang.Thread.run(Thread.java:833)
> Caused by: jakarta.transaction.RollbackException: ARJUNA016053: Could not commit transaction.
>         at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1283)
>         at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
>         at io.quarkus.narayana.jta.runtime.NotifyingTransactionManager.commit(NotifyingTransactionManager.java:70)
>         at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.endTransaction(TransactionalInterceptorBase.java:406)
>         at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:171)
>         at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
>         at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
>         at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
>         at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
>         at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)
>         at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
>         at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)
>         at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
>         at io.quarkus.resteasy.reactive.server.runtime.StandardSecurityCheckInterceptor.intercept(StandardSecurityCheckInterceptor.java:44)
>         at io.quarkus.resteasy.reactive.server.runtime.StandardSecurityCheckInterceptor_RolesAllowedInterceptor_Bean.intercept(Unknown Source)                                                                                                                                            
>         at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
>         at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
>         at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
>         ... 13 more                                                   
> Caused by: jakarta.validation.ConstraintViolationException: Validation failed for classes [com.redhat.swatch.contract.repository.ContractEntity] during persist time for groups [jakarta.validation.groups.Default, ]
> List of constraint violations:[                                       
>         ConstraintViolationImpl{interpolatedMessage='must not be empty', propertyPath=metrics, rootBeanClass=class com.redhat.swatch.contract.repository.ContractEntity, messageTemplate='{jakarta.validation.constraints.NotEmpty.message}'}
> ]                                                                     
>         at org.hibernate.boot.beanvalidation.BeanValidationEventListener.validate(BeanValidationEventListener.java:151)
>         at org.hibernate.boot.beanvalidation.BeanValidationEventListener.onPreInsert(BeanValidationEventListener.java:81)
>         at org.hibernate.action.internal.EntityInsertAction.preInsert(EntityInsertAction.java:215)
>         at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:98)
>         at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:635)
>         at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:502)
>         at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:364)
>         at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
>         at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
>         at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1412)
>         at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:485)
>         at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2301)
>         at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1966)
>         at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439)
>         at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:336)                                                                                                                                 
>         at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
>         at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
>         at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:52)
>         at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:351)
>         at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:69)
>         at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:138)
>         at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1271)
>         ... 30 more                                                   
> ```

This is due to the latest rebase. I've already fixed this to discard the wrong metrics before our validation.

The question now is if a contract with no metrics is valid or not. Any ideas? @kahowell @awood

Sgitario commented on swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java:
```diff
@@ -90,63 +83,45 @@ public enum ContractMessageProcessingResult {
       MeasurementMetricIdTransformer measurementMetricIdTransformer,
-      SubscriptionSyncService syncService) {
+      SubscriptionSyncService syncService,
+      AwsPartnerEntitlementsProvider awsPartnerEntitlementContract,
```
Ops, good catch!

Sgitario replied:
PR updated.

ntkathole commented:
/retest

ntkathole commented:
QE : check contract tests failures - the API of contract endpoint changed- https://github.com/RedHatInsights/rhsm-subscriptions/blob/00466d7b1b57de7b0de08a9232cfd2cf2f9f10fc/swatch-contracts/README.md#mapping-partner-api-with-swatch-data

kartikshahc commented:
> > And coincidentally enough, when I make the `curl` call listed, the following stack gets thrown
> > ```
> > .quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method                                                                                                                                                                                                              
> >         at com.redhat.swatch.contract.resource.ContractsTestingResource_Subclass.createContract(Unknown Source)                                                                                                                                                                           
> >         at com.redhat.swatch.contract.resource.ContractsTestingResource_ClientProxy.createContract(Unknown Source)                                                                                                                                                                                                         
> > ```
> 
> This is due to the latest rebase. I've already fixed this to discard the wrong metrics before our validation and also updated the test steps from the PR description.
> 
> The question now is if a contract with no metrics is valid or not. Any ideas? @kahowell @awood

I think contract with no metrics is valid based on the recent discussions we were having. Correct @kahowell ?

kahowell commented:
> I think contract with no metrics is valid based on the recent discussions we were having. Correct @kahowell ?

Correct, a contract with no metrics is valid.

kahowell reviewed (changes requested):
Need to make metrics optional, previously we treated it valid if there was at least one metric in the message or API response, but this is inconsistent of actual final state of a contract record, where it is valid to have no metrics at all.

Sgitario commented:
> Need to make metrics optional, previously we treated it valid if there was at least one metric in the message or API response, but this is inconsistent of actual final state of a contract record, where it is valid to have no metrics at all.

Ok, I will remove the NotEmpty annotation and update the PR.

Sgitario commented:
@kahowell @kartikshahc PR updated. Thanks for your comments.

Aurobinda55 commented:
/retest

Aurobinda55 commented:
/retest

Aurobinda55 commented:
`Verification steps `

- From Iqe terminal , Create an  aws contract without any metrics 

```
 contract_data={
   ...:  'billing_account_id': 'test129',
   ...:  'sku': 'MW02393',
   ...: 'start_date': '2024-02-01T10:44:55Z'}

application.rhsm_subscriptions.create_contract(product_id="rosa" ,**contract_data)`

2024-02-27 08:50:43.581 [    INFO] [root] Response: {
    "contract": {
        "billing_account_id": "test129",
        "billing_provider": "aws",
        "billing_provider_id": "a3061b2a-94db-434e-903b-80e4999bb6a2;aws_customer_id_123;45612",
        "end_date": "2024-02-29T10:44:55Z",
        "metrics": [],
        "org_id": "14541308",
        "product_id": "rosa",
        "sku": "MW02393",
        "start_date": "2024-02-01T10:44:55Z",
        "subscription_number": "4327953",
        "uuid": "fb955652-590f-496d-b7e9-d838c491dd07",
        "vendor_product_code": "f8ebd078-4ee8-4a01-8caa-6afcbaa244f2"
    },
    "status": {
        "message": "New contract created",
        "status": "SUCCESS"
    }
}

```
- Again create an azure contract without any metrics 

```
contract_data1={
   ...:  'billing_account_id': 'test135',
   ...:  'sku': 'RH02781HR',
   ...: 'start_date': '2024-02-05T10:44:55Z'}

In [11]: application.rhsm_subscriptions.create_contract(product_id="rhel-for-x86-els-payg" ,**contract_data1)

2024-02-27 09:16:08.120 [    INFO] [root] Response: {
    "contract": {
        "billing_account_id": "test135;test135",
        "billing_provider": "azure",
        "billing_provider_id": "azure_record_id_123;plan_id_1234;746a5243-2f58-4072-9b19-39bbdbdfb3ea",
        "end_date": "2024-02-29T10:44:55Z",
        "metrics": [],
        "org_id": "14541308",
        "product_id": "rhel-for-x86-els-payg",
        "sku": "RH02781HR",
        "start_date": "2024-02-05T10:44:55Z",
        "subscription_number": "1372714",
        "uuid": "72c8bb1b-db5e-4e3f-8a66-987b130bb210",
        "vendor_product_code": "0ac8231b-1774-4392-8414-fd002b6f0acf"
    },
    "status": {
        "message": "New contract created",
        "status": "SUCCESS"
    }
}

```

- Create aws contract with valid  metrics

```
contract_data1={'instance_hours': 5,
   ...:  'cores': 7,
   ...:  'billing_provider': 'aws',
   ...:  'billing_account_id': 'test140',
   ...:  'sku': 'MW02393',
   ...: 'start_date': '2024-02-05T10:44:55Z'}

In [7]: application.rhsm_subscriptions.create_contract(product_id="rosa" ,**contract_data1)

Out[7]: 
{'status': {'status': 'SUCCESS', 'message': 'New contract created'},
 'contract': {'uuid': 'ba76edd4-0bbe-47b4-b0a6-d8d48945b3ae',
  'subscription_number': '3535449',
  'sku': 'MW02393',
  'start_date': '2024-02-05T10:44:55Z',
  'end_date': '2024-02-29T10:44:55Z',
  'org_id': '14541308',
  'billing_provider': 'aws',
  'billing_provider_id': '631b0153-d719-4f3b-8236-62c9722fd8ce;aws_customer_id_123;45612',
  'billing_account_id': 'test140',
  'product_id': 'rosa',
  'vendor_product_code': 'ced373ac-c155-4826-ad4a-6f595ecf3abc',
  'metrics': [{'metric_id': 'control_plane', 'value': 5},
   {'metric_id': 'four_vcpu_hour', 'value': 7}]}}
``` 

- Verified a SUCCESS status message when the contract created successfully 
- Verified  a contract created successfully without any metrics 
- Verified the start and end date 
- Verified the billing_provider_id is as per the design document for aws and Azure 
- Verified the plan id filed in Azure contract

Aurobinda55 commented:
@Sgitario I have manually verified the changes and everything looks good, also raised a PR for necessary changes in smqe tools, will update the status of the failed tests once the PR approved and merged  . Thanks

Sgitario commented:
> @Sgitario I have manually verified the changes and everything looks good, also raised a PR for necessary changes in smqe tools, will update the status of the failed tests once the PR approved and merged . Thanks

Many thanks!

Sgitario commented:
> @Sgitario I have manually verified the changes and everything looks good, also raised a PR for necessary changes in smqe tools, will update the status of the failed tests once the PR approved and merged . Thanks

Any updates on this @Aurobinda55 ?

Aurobinda55 commented:
/retest

Aurobinda55 commented:
> > @Sgitario I have manually verified the changes and everything looks good, also raised a PR for necessary changes in smqe tools, will update the status of the failed tests once the PR approved and merged . Thanks
> 
> Any updates on this @Aurobinda55 ?

@Sgitario I got some review comments on my PR, will try to finish it today  and update you soon

Aurobinda55 commented:
/retest

Aurobinda55 commented:
/retest

Aurobinda55 commented:
> > > @Sgitario I have manually verified the changes and everything looks good, also raised a PR for necessary changes in smqe tools, will update the status of the failed tests once the PR approved and merged . Thanks
> > 
> > 
> > Any updates on this @Aurobinda55 ?
> 
> @Sgitario I got some review comments on my PR, will try to finish it today and update you soon

@Sgitario QE pr -https://gitlab.cee.redhat.com/smqe/smqe-tools/-/merge_requests/256 is merged now, but I see build failures now on your PR might be due to the merge conflict issue, will fix the failed tests once I got a successful build.

Sgitario commented:
> > > > @Sgitario I have manually verified the changes and everything looks good, also raised a PR for necessary changes in smqe tools, will update the status of the failed tests once the PR approved and merged . Thanks
> > > 
> > > 
> > > Any updates on this @Aurobinda55 ?
> > 
> > 
> > @Sgitario I got some review comments on my PR, will try to finish it today and update you soon
> 
> @Sgitario QE pr -https://gitlab.cee.redhat.com/smqe/smqe-tools/-/merge_requests/256 is merged now, but I see build failures now on your PR might be due to the merge conflict issue, will fix the failed tests once I got a successful build.

yes... actually, the changes in https://github.com/RedHatInsights/rhsm-subscriptions/pull/3089 that caused the conflicts are incompatible with these changes. We need to hold on these changes until we agree with how to continue. I'll schedule a meeting to discuss it. In the mean time, since we're not merging these changes, perhaps do you need to revert the changes from the MR that you has just merged?

Sgitario commented:
@Aurobinda55 @kahowell I've already updated the PR to not use the message from the Gateway but only the request from the Partner API. 
@Aurobinda55 I've also updated the PR instructions and the README.md, the request to be provided is less complicated now but you will need to update the MR of the sqme-tools repo.

Sgitario commented:
/retest

ntkathole commented:
/retest

Aurobinda55 commented:
> @Aurobinda55 @kahowell I've already updated the PR to not use the message from the Gateway but only the request from the Partner API. @Aurobinda55 I've also updated the PR instructions and the README.md, the request to be provided is less complicated now but you will need to update the MR of the sqme-tools repo.

@Sgitario working on this , will update soon

Aurobinda55 commented:
re-verified the above verification steps and looks good to me, created two PRs to make the tests green.

ntkathole commented:
/retest

Aurobinda55 commented:
/retest

ntkathole commented:
/retest
